### ISSUE TITLE
Exporter wishlist

### ISSUE THREAD

**Issue by oantolin** [oantolin/embark#95]
I think I might be missing many opportunities for export. Let's list some here in this issue:

- `bookmark` should export to bookmark-bmenu-mode.
- `line` should export to occur-mode.

**Comment by oantolin** [comment]
We should also think about exporters run from non-mini buffers. One that has been suggested is that in an eww buffer, export could make a list of all the links. I'm not so sure about that, probably a `consult-link` command is better and you can always run embark export from inside that hypothetical `consult-link`.

But I'm sure there are some good opportunities for export from regular buffers I'm missing. Currently the only ones are kind of boring: dired buffers export their files, ibuffers export their buffers, `*Completions*` buffers export their completions (you can also act on completions in a `*Completions*` buffer, so this export is kind of pointless --although I guess you do gain grid and zebra views, and the direct action minor mode).

**Comment by minad** [comment]
Is there also a target finder wishlist? Or is the list already exhautive?

**Comment by oantolin** [comment]
The target finder list is not exhaustive at all! Should I make a separate issue for that or just broaden this one?

**Comment by minad** [comment]
I would use one wishlist as I have in consult. I think this works quite well as some kind of general discussion thread before the details are then discussed in a separate dedicated issue.

**Comment by oantolin** [comment]
Earlier when I mentioned eww links I was really talking about just collecting them and forgot to say what kind of buffer they should export to: maybe an `eww-bookmark-mode` buffer is reasonable, even though they aren't bookmarks.

**Comment by oantolin** [comment]
OK, I've broadened this issue on your advice @minad. (I don't want the wishlist to also cover new commands, since in fact, I'm pretty skeptical about adding any new top-level commands to Embark at this point, except possibly the resume functionality, if we think it fits better in Embark.)

**Comment by oantolin** [comment]
Oh, I just remembered some actions for variables I often miss: `set-variable` and `customize-set-variable`, maybe even `customize-variable`.

**Comment by oantolin** [comment]
Good news: now `consult-line`, `consult-mark` and `consult-global-mark` export to occur mode buffers and `e` (for `occur-edit-mode`) works in them!

**Comment by minad** [comment]
@oantolin This is amazing! I already have it mentioned here: https://github.com/minad/consult#integration-with-embark. Now I can remove the work in progress note :partying_face:

**Comment by oantolin** [comment]
> Oh, I just remembered some actions for variables I often miss: `set-variable` and `customize-set-variable`, maybe even `customize-variable`.

I added these for variables, and for commands I added `execute-extended-command`, `global-set-key` and `local-set-key`. 846cfbd7212ba3a3a47a059ba35400d534968815

**Comment by protesilaos** [comment]
I am just sharing this good idea for targeting video links: https://jao.io/blog/2021-01-09-embarking-videos.html. Natural complements are audio files and an action to yt-dl the target (audio could call `mpv` with the `--no-video` flag).

Maybe this is something that can be of use.

**Comment by sheijk** [comment]
@oantolin the actions for variables and commands are not available here. I checked that the commit you referenced is applied, value of embark-keymap-alist set accordingly, etc. But when I act on something like `(defcustom e|mbark-keymap-alist` (point at "|" char) I don't see customize or set variable options, same with commands like embark-act (version 20210111.1408). Can I see which map has been selected somewhere?

**Comment by sheijk** [comment]
org-mode classifiers and actions would also be great. Or would this be better for an external package?

**Comment by oantolin** [comment]
@sheijk The variable and command specific actions are currently only available from minibuffer commands for those categories. For example you get the command actions for candidates in M-x in recent Emacsen. If you use Marginalia, then you get the command actions in M-x even on older Emacsen. And again, with Marginalia, you get the variable actions from describe-variable, for example, or from customize-variable.

The point is that, as you found, they are not supposed to work on symbols in regular buffers... yet! I've been thinking about refining the type. It would be much better once the multiple-targets stuff is finished #92, but for now I guess I could just pick an order: prefer command over variable if the symbol is both things. If Emacs were extended in Guile we wouldn't have these problems... 😛

**Comment by oantolin** [comment]
> org-mode classifiers and actions would also be great. Or would this be better for an external package?

Possibly a separate package would be better, I'm not sure. I mean Org does come with Emacs. But that can be decided later, the first thing is to decide what kinds of targets and actions would be useful in Org buffers. Did you have any thing in mind? For a target type to be really useful ideally it should have several associated actions.

**Comment by sheijk** [comment]
Separate package makes sense, or at least a separate elisp file to not require loading org-mode when embark gets used in other buffers. I've recently made a little [transient based menu](https://github.com/sheijk/org-menu) for org, I think most actions offered there would be good embark candidates. Some examples for actions would be:

- TODO items - toggle next state, set explicit state, remove state
- Headline - same as TODO, org-update-checkbox-count, move up/down, (de)indent, cut tree
- Timestap - change date, toggle type, show in calendar
- Link - open, edit
- Region - toggle bold/italic/..
- List items - set to "[ ]", "[x]", "[-]", no checkbox, indent/outdent, org-list-repair, org-list-make-subtree, cycle-style, maybe move up/down
- Table: evaluate (cell/line/all/all-until-nothing-changes), add/delete/move row/column, edit field, blank field, org-copy-down, set field/column formula

**Comment by oantolin** [comment]
I think your suggested org targets and actions are very good and that a separate `embark-org` (or `org-embark`?, how does that work?) package with them would be reasonable, but, wow, your org-menu package looks fantastic, @sheijk! I'd say your package takes the pressure off Embark to provide similar functionality with a different interface. 😄

**Comment by minad** [comment]
You could create an embark-org.el file similar to consult-flymake.el in order to retain lazy loading.

**Comment by sheijk** [comment]
@oantolin or maybe start it with the actions I already made. Are you looking for contributions (no promise!)?

**Comment by oantolin** [comment]
@sheijk I added a transformer to refine `symbol` to `command` or `variable` if possible, for when you act on the symbol at point in a regular buffer. 296ad8077092ff839029c0bdb53c279e921e5354

**Comment by oantolin** [comment]
> I am just sharing this good idea for targeting video links: https://jao.io/blog/2021-01-09-embarking-videos.html. Natural complements are audio files and an action to yt-dl the target (audio could call `mpv` with the `--no-video` flag).

Thanks for pointing this out, @protesilaos. I wrote to jao to suggest a slightly different implementation [which he wrote about](https://jao.io/blog/2021-01-11-an-even-better-video-wharf.html), and I noticed he also wrote an [async Spotify client using Consult and Marginalia](https://jao.io/blog/2021-01-08-espotify.html)!

**Comment by protesilaos** [comment]
> oantolin: I wrote to jao to suggest a slightly different implementation which he wrote about, and I noticed he also wrote an async Spotify client using Consult and Marginalia!

Yes, I have read them.  My plan is to copy and adapt the video targeting part (never used spotify).  More generally, I am excited to see what kind of solutions, tips and tricks the community comes up with for those packages.

**Comment by minad** [comment]
@oantolin @protesilaos Regarding the spotify client, it is exciting that they implemented a web access using the async api! I have some small experiments in the consult websearch branches as well, but they turned out to be not useful.

**Comment by astoff** [comment]
I think the way Embark works on a regular buffer is too Elisp-centric.  Hopefully most people spend most of their time in Emacs doing other things!

For the purposes of `embark-target-symbol-at-point`, I would argue that symbol should mean a symbol in the current major mode.  So the `d` action would be bound to `xref-find-definitions`.  It's less clear what to bind to `h`, since there's no “describe thing at point” in xref (or whatever Emacs component this would fit into).  Anyway, I suggest binding the `h` key to `display-local-help`.  This is mostly useless in itself, but at least the user can remap that command to suitable major-mode specific values (`describe-symbol` in Elisp, `cider-doc` in Cider, etc.).  Plus, it would work out-of-the-box with Eglot, which does precisely that remapping in its own mode-map.

On the other hand, I'm fine with the `eval-region` binding in `embark-region-map` because there's no better option, and, as above, one can use remaps to improvise a generic interface to the “send to repl” action across major modes.

There could still be an `embark-target-emacs-lisp-symbol-at-point`, but it should probably check if the buffer is in Elisp mode.  Then it wouldn't work in random places like help buffers, but at least Embark would not capriciously (not) recognize words in regular text or other programming languages just because they happen to (not) have a meaning in Elisp.

**Comment by minad** [comment]
@astoff I agree! But I guess thing-at-point is also elisp centric. Unfortunately elisp is somehow more first class in Emacs then the other major modes. But that is natural since you can see it as a whole programmable environment.

**Comment by astoff** [comment]
@minad `thing-at-point` uses the syntax table to decide what is a symbol. (So in most prog-modes, but not in lisp-mode, (thing-at-point 'symbol) just after the text `x-y` returns `y`.)

**Comment by minad** [comment]
@astoff Thanks! Good to know!

**Comment by sheijk** [comment]
Another suggestion: support find-file-at-point for embark-become

**Comment by oantolin** [comment]
Those are all very good points, @astoff. I think the symbol stuff is so Emacs Lisp centric because I used it to develop it 😛

I've been thinking that the symbol target finder shouldn't check to see if the symbol is bound anyway. It uses `(thing-at-point 'symbol)` so it already takes syntax into account. I'll just remove the check for bindings from the target finder.

As for the actions, I use it all the time to lookup help or definitions for Emacs Lisp symbols in non-Emacs Lisp buffers (mostly help, markdown, org and message buffers) and I'd really, really hate to do lose that. I guess I could make the configuration that ships with Embark more "politically correct" along the lines you suggest, and put a proper Emacs Lisp-centric configuration in my personal init.el. 😛 

Or maybe I could still bind those Emacs Lisp centric commands, to different keys. I can't be the only one who wants to just to the definition of an Emacs Lisp symbol in a help buffer and is frustrated that `xref-find-definitions` prompts you for a tags table.

**Comment by minad** [comment]
> xref-find-definitions prompts you for a tags table.

You can fix that by disabling the obsolete etags backend.

**Comment by oantolin** [comment]
> Another suggestion: support find-file-at-point for embark-become

I don't understand this suggestion, @sheijk. Do you mean you want `find-file-at-point` bound in the `embark-file+buffer-map`? I don't think it would be very useful: `find-file-at-point` is basically the same as `find-file`, except the default is the file at point. As soon as you type something at the prompt, which embark-become does for you, it becomes just like find-file.

Could you explain what you meant.

**Comment by oantolin** [comment]
> You can fix that by disabling the obsolete etags backend.

I'm confused: disabling that backend makes xref think that help buffer contain Emacs Lisp symbols?

**Comment by sheijk** [comment]
If I call `find-file` I can move to other file contexts. But if I call `find-file-at-point` embark-become doesn't offer any targets to jump to

**Comment by astoff** [comment]
From Elisp buffer, does xref work for you, @oantolin? `xref-find-definitions` has the additional benefit that you can go back to where you were before with `xref-pop-marker-stack`. Those things also have nice and prominent keybindings anyway, but it would be nice for the sake of consistency if Embark knew about them.

> I use it all the time to lookup help or definitions for Emacs Lisp symbols in non-Emacs Lisp buffers

Fair enough. You could recover this behavior in one of two ways: (1) remapping `xref-find-definitions` and `display-local-help` in enough modes, or (2) introducing `embark-target-emacs-lisp-symbol-at-point`, which could or could not be on by default, and could perhaps also check if the major mode derives from one in a customizable list.

**Comment by oantolin** [comment]
> If I call `find-file` I can move to other file contexts. But if I call `find-file-at-point` embark-become doesn't offer any targets to jump to

Gotcha! You don't want it so other commands become it but for it to become other commands. That makes perfect sense, sorry I didn't see that at once. I added it. 9fa6d6adc0086a97d82bf36723d44490bb6c9aa1

**Comment by oantolin** [comment]
> From Elisp buffer, does xref work for you, @oantolin 

Yes, of course. I use it there instead of Embark actions so I can use `xref-pop-marker-stack`. I use the Embark actions on Emacs Lisp symbols in a variety of non-Emacs Lisp buffers.

> > I use it all the time to lookup help or definitions for Emacs Lisp symbols in non-Emacs Lisp buffers
> 
> Fair enough. You could recover this behavior in one of two ways: (1) remapping `xref-find-definitions` and `display-local-help` in enough modes, or (2) introducing `embark-target-emacs-lisp-symbol-at-point`, which could or could not be on by default, and could perhaps also check if the major mode derives from one in a customizable list.

Or (3), having both `xref-find-defintions` and `embark-find-definition` bound to different keys in the symbol action map. They do, after all, do different things. And I have wanted both in the same buffer at times.

I wouldn't really want a separate `embark-target-emacs-lisp-symbol-at-point`, specially one that's not on by default. Does really no-one else want Emacs Lisp symbol actions to be on by default in `*Help*` buffers? I'd find that very hard to believe.

**Comment by minad** [comment]
> Does really no-one else want Emacs Lisp symbol actions to be on by default in *Help* buffers? I'd find that very hard to believe.

:+1:

Would it make sense to add a possibility to select keymaps based on some predicate? This way you could exchange keymaps depending on the major mode for example.

**Comment by oantolin** [comment]
> Would it make sense to add a possibility to select keymaps based on some predicate?

Seems unnecessary, just use a transformer to change the type based on the major mode. Something like:

```elisp

(defun refine-symbol-type-by-major-mode (target)
  (cons (or (alist-get major-mode '((java-mode . java-symbol)
                                    (cc-mode . c-symbol)))
            'symbol)
        target)
```

And associate `java-symbol`, etc. to whatever action map you want.

**Comment by astoff** [comment]
>  Does really no-one else want Emacs Lisp symbol actions to be on by default in Help buffers? I'd find that very hard to believe.

Me too. And maybe this is a bug in xref! But then, as you mentioned, there are other places beyond Help buffers where it's perfectly reasonable to want actions on Elisp symbols. I'm not against special treatment to Elisp, but I'm arguing for staying as generic as possible, given certain limitations of Emacs itself.

> Or (3), having both `xref-find-defintions` and `embark-find-definition` bound to different keys in the symbol action map.

I wouldn't really want _that_, I'm afraid... the xref stuff already has nice keybindings. I think the interesting thing here is to make the action context-dependent. So just one key for “show documentation”, and it does what you mean everywhere.

We're discussing the tail of `embark-target-finders` here, and maybe this is not even a super interesting case, so it shouldn't be overengineered :-). Don't you think a `embark-target-emacs-lisp-symbol-at-point`, placed as the next-to-last` finder, and of the shape
```
(when (apply 'derived-mode-p 'embark-modes-where-we-care-about-stray-elisp-symbols)
  ...)
```
would be reasonable? The very last finder would then be the one for symbols as per the major mode, which relies on xref (plus `display-local-help` on `h` as long as xref doesn't offer anything suitable).

**Comment by oantolin** [comment]
Do you want  a separate finder for Emacs Lisp symbols because you want it always recognize Emacs Lisp symbol syntax independent of the major mode syntax for symbols? I've been just using `(thing-at-point 'symbol)` in all major modes.

> The very last finder would then be the one for symbols as per the major mode, which relies on xref (plus display-local-help on h as long as xref doesn't offer anything suitable).

Just to make sure I understand you: you don't literally mean the *finder* would depend on xref, right? The finder would use `(thing-at-point 'symbol)`, and what you mean is that the actions offered would include xref actions?

**Comment by astoff** [comment]
> Do you want a separate finder for Emacs Lisp symbols because you want it always recognize Emacs Lisp symbol syntax independent of the major mode syntax for symbols?

No, all that I care about is for `M-x embark-act RET h` to do the right thing on any buffer. It doesn't on a Python buffer. And just to clarify, I can easily do this for myself, I was just wondering if there's a nice, generic, unsurprising approach that requires no configuration. Seems quite tricky, though.

> Just to make sure I understand you: you don't literally mean the _finder_ would depend on xref, right?

Yes, right, maybe I put it a bit confusingly. I was imagining separate keymaps for Elisp symbols and for "symbols according to the major mode". This could be achieved through separate finders, but not only that way, as you pointed out above.

**Comment by oantolin** [comment]
OK, I think I understand and agree now. What actions are appropriate for non-Emacs Lisp symbols besides `xref-find-definitions` and `display-local-help`? Should they only get those two (plus the actions in embark-general-map, of course).

**Comment by astoff** [comment]
> What actions are appropriate for non-Emacs Lisp symbols besides xref-find-definitions and display-local-help?

That's all there is, I think.

**Comment by oantolin** [comment]
OK. Got it. Will do. I'll call that new non-Emacs Lisp symbol type `identifier`, which is the term `xref-find-definitions` uses.

**Comment by astoff** [comment]
Good, now that I think about it, "identifier" is really the standard terminology. Symbols go by the weird name of "string" in most languages these days :rofl:

**Comment by oantolin** [comment]
OK, @astoff. I made the change. Now `(thing-at-point 'symbol)` is always recognized as a valid target and it is classified as either `symbol` (meaning Emacs Lisp symbol), or as `identifier`. The ones classified as symbols are those _either_ were found in an Emacs Lisp buffer (whether or not they are bound), _or_ are interned in the Emacs Lisp environment and were found in a buffer whose major mode does not inherit from `prog-mode`. Does that sound OK? beb5dd604890f4920e39bebefdd5619c81626fdc

**Comment by oantolin** [comment]
For identifiers the only actions you get are, as you suggested, `xref-find-definitions` and that lovely function `display-local-help` which is used to print the message "No local help at point". 😛

**Comment by TxGVNN** [comment]
Hello everyone. I'm thinking about adding an action `run M-x` when I run `find-file`

With this action I can :
- Run `find-grep`/`rg` with pointed directory
- Run `eshell/shell/ansi-term/vterm` under pointed directory
- Any commands require default-directory

WDYT?

**Comment by oantolin** [comment]
Do you mean you want to be able to run `find-file`, maybe navigate to some directory, and then run `embark-act` but instead of choosing one of the actions in the file map, use `M-x` to run whatever command you wan. @TxGVNN? If that's what you mean you can already do it!

If you use the keymap prompter, just type `M-x` and the command; if you use the completing-read prompter, when it prompts you for a command just type the command you want, it does *not* need to be one the list, those are just suggestions.

**Comment by TxGVNN** [comment]
Thank @oantolin, That is possible, right. I just suggest adding this action to the package if you think it is useful for everyone. I will build it for me.

**Comment by oantolin** [comment]
Maybe I don't understand what your action does, does it just prompt you for a `M-x` command and then run it? If so, I believe that action in completely unnecessary, since `embark-act` already lets you use `M-x` to run an arbitrary command. What advantage would your `Run M-x` action have?

**Comment by oantolin** [comment]
> Thank @oantolin, That is possible, right. I just suggest adding this action to the package if you think it is useful for everyone. I will build it for me.

What I am trying to say is that *it has already been added to the package* and that *you do not need to build it*, since it already exists, @TxGVNN. Or am I misundertanding what your action would do?

**Comment by oantolin** [comment]
Oh, maybe I misunderstood you, @TxGVNN! Do you want the action to use the target just to set the default directory and **not** to pass the target to the command you run? Like this?

```elisp
(defun embark-in-directory (dir cmd)
  "Run CMD in directory DIR."
  (interactive "DIn directory: \nCRun command: ")
  (let ((default-directory dir))
    (execute-command cmd)))
```

**Comment by TxGVNN** [comment]
Oh, yes! That is my idea. 

```emacs-lisp
(defun embark-in-directory (dir cmd)
  "Run CMD in directory DIR."
  (interactive "DIn directory: \nCRun command: ")
  (let ((default-directory (file-name-directory dir)))
    (command-execute cmd)))

(define-key embark-file-map (kbd "M-x") #'embark-in-directory)
```

**Comment by oantolin** [comment]
Ah, good! Sorry for the initial misunderstanding.

By the way, I would *not* recommend the `M-x` binding for that, because you'd loose the ability to use `M-x` to pick an action (unless you happen to have a different key binding for `execute-extended-command`). Maybe `/`, because the action is used to pick a directory?

**Comment by TxGVNN** [comment]
Yup. I'm binding `M-o` to `embark-act`. But if you add this action, I see `/` is a good default keybinding.

**Comment by astoff** [comment]
@TxGVNN I feel like this is not actually an Embark matter (unless we're discussing the need of an Embark wrapper for an already existing functionality).

If you want a command to execute other commands in a different directory, how about something like this?

``` elisp
(defun mirror-command-execute (command)
  "Execute COMMAND with remote `default-directory'.
Interactively, reads a key sequence, so it acts like a keymap."
  (interactive (list (lookup-key (current-active-maps)
                                 (read-key-sequence "On remote: "))))
  (unless (commandp command)
      (error "`%s' is not a valid command" command))
  (setq this-command command)
  (let ((default-directory mirror-location))
    (command-execute command)))
```
When called interactively, this reads the command not from a `M-x` style prompt, but from a key sequence, so it acts like a fake keymap. Of course you can always type `M-x` to get a prompt.

Also, this sets the `default-directory` to the value of a directory-local variable `mirror-location`, since, in my use case, the “remote” directory is fixed, on a per-project-basis. Of course you can change that and completing-read a directory name instead. (The version I actually use computes the remote directory based on the buffer's file name, the local project root, and remote project root. But that's another story.)

**Comment by TxGVNN** [comment]
Thank @astoff, I'm glad with `embark-in-directory`action right now. I love it. I had a same feature with ivy action in past.

If I want to grep something in a other directory, I always find-file then run grep on it. That is my workflow.
I have 3 scopes to run a command:
- current dir scope -> it is default in all commands. Example: `consult-grep`
- project scope -> I will run with projectile-* command. Example: `consult-projectile-grep` (my function)
- other dir scope -> I use find-file then action with `embark-in-directory`

**Comment by minad** [comment]
The embark-in-directory action is a nice idea :)

**Comment by sheijk** [comment]
`kill-buffer` support would be great. Ideally without closing the completion afterwards

**Comment by minad** [comment]
@sheijk kill-buffer is already part of the buffer actions (bound to k), but it is affected by #117 too, when keeping the completion system open.

**Comment by oantolin** [comment]
> kill-buffer support would be great

@sheijk You can run `kill-buffer` as an action in any command that offers completion of buffer names, such as `switch-to-buffer` or `kill-buffer` itself. As @minad said, it is bound to `k`.

> Ideally without closing the completion afterwards so multiple buffers can be killed on-the-fly while switching between them

The `embark-act` command can be run in two ways: so that it quits the minibuffer after you run or in a way that keeps the minibuffer open. This is controlled by the variable `embark-quit-after-action` which is `t` by default, which means `embark-act` quits the minibuffer and `C-u embark-act` does not. If you set `embark-quit-after-action` to `nil` (which I prefer) it's the other way around, `embark-act` keeps the minibuffer open and `C-u embark-act` quits.

I often do exactly what you asked, @sheijk, kill several buffers without quitting a `switch-to-buffer` command.

@minad said:

> but it is affected by #117 too, when keeping the completion system open.

The built-in buffer commands are careful to make sure the buffers in the candidate lists are still live, so for these commands simply refreshing the completion UI is enough. The refresh can be triggered manually (by adding and removing a character) or automatically with [the configuration from the wiki](https://github.com/oantolin/embark/wiki/Additional-Configuration#refresh-candidate-list-after-action).

**Comment by minad** [comment]
@oantolin 

> The built-in buffer commands are careful to make sure the buffers in the candidate lists are still live, so for these commands simply refreshing the completion UI is enough. The refresh can be triggered manually (by adding and removing a character) or automatically with the configuration from the wiki.

Are you sure? switch-to-buffer+kill does not for me with Selectrum. Since you moved the Selectrum config to embark.el, I am using a barebone Embark configuration without configuring any hooks myself besides the which-key indicator.

**Comment by oantolin** [comment]
Ah, when will I learn to always test with Selectrum? It works with default completion, icomplete and embark collect live or completions. Selectrum is more static, I forget it only gets the list of candidates once. I don't even know why we have a refresh snippet for it on the wiki. Does the snippet actually do anything?

**Comment by minad** [comment]
No idea. I hope the non-functional snippet is not part of embark.el :smile: While the way Selectrum does things is a bit non-standard, I think it is still valid. Selectrum applies some kind of normalization first to get all the candidates. This is a perfectly valid thing given how baroque the completing-read API is. Maybe `switch-to-buffer` would work if you use `selectrum-exhibit` to update, since this works with Consult async and ensures that the candidates are reread. However I still think we need #117, since for example `consult-buffer` would not update without it. And I don't see how I could possibly change `consult-buffer`/`consult--multi` to magically pick up changes to the original candidate lists.

**Comment by oantolin** [comment]
The snippet is not in embark.el, just on the wiki. And it used to work with `switch-to-buffer`, too. Now it doesn't seem to, even adding `(selectrum-exhibit)`...

**Comment by oantolin** [comment]
The snippet is not in embark.el (which indeed, does not try to automatically refresh completion UIs at all), just on the wiki. And I remember it used to work with `switch-to-buffer`. It doesn't seem to work now, even if you add a post action hook that calls `selectrum-exhibit`.

@clemera, is this out of date?

```elisp
(defun refresh-selectrum ()
  (setq selectrum--previous-input-string nil))

(add-hook 'embark-pre-action-hook #'refresh-selectrum)
```

I remember that it used to removed buffers I just killed from `switch-to-buffer`.

**Comment by clemera** [comment]
To refresh you should better call `selectrum-exhibit` now. BTW I just noticed weird issues which seems to be introduced by  e20a1ec093cdf8877a93511277208794fea03ce0. When using embark keymap help and prescient sorting candidates get inserted into other commands afterwards by `selectrum-prescient--preprocess` :eyes:  I don't know what is happening yet.

**Comment by clemera** [comment]
Additionally I get weird display issues when using it multiple times these go away when removing the recent change to highlighting in Selectrum (you get the other display issues back in exchange) but the inserting of embark keymap help into other commands persists.

**Comment by oantolin** [comment]
That's really weird, @clemera. embark-completing-read-prompter just calls completing-read as it did before, I don't see why this change would make any difference in prescient.

**Comment by clemera** [comment]
It doesn't happen before e20a1ec093cdf8877a93511277208794fea03ce0 but I also don't know what is happening. I'm currently trying to figure out the issue with regard to the new display issue first.

**Comment by clemera** [comment]
This display issue is really strange, too. The `(symbol-name cmd)` in `embark-completing-read-prompter` returns a string which already has the `display` property applied, seems like the string is cached in memory and reused or something like that?

**Comment by oantolin** [comment]
Ah, I thought that was always a fresh string. I'll copy it before adding the display attributes.

**Comment by clemera** [comment]
Yes, using `substring-no-properties` on it seems to fix it.

**Comment by oantolin** [comment]
Oh, I used `concat` instead. But I think it should work too? 3ccb9c2721f9bee854fe50be50d9724953bc5556

**Comment by clemera** [comment]
I think that should work, too I will check.

**Comment by clemera** [comment]
Yes, that works, thanks!

**Comment by clemera** [comment]
This also fixes the other issue, apparently the cached string with the display property was then also later reused when using `M-x` which would display `embark-insert` with the prefix :laughing:

**Comment by clemera** [comment]
I did not notice it without prescient because then they wouldn't be sorted first...

**Comment by oantolin** [comment]
> This also fixes the other issue, apparently the cached string with the display property was then also later reused when using `M-x` which would display `embark-insert` (or other candidates of the completion help) with the prefix 😆

Ah, OK! Great, @clemera!

**Comment by oantolin** [comment]
Add an exporter for bookmarks. e361607d448243b1a5c6b6eac71627e7636fdabc

**Comment by sheijk** [comment]
A detector for flyspell errors (face flyspell-incorrect) and an action to call ispell-word (M-$ by default bindings) on it. A detector for flymake/flyspell would also be nice altough users would need to provide actions. Maybe "copy error message" and "copy error location"?

**Comment by oantolin** [comment]
Over on minad/consult#210, @hmelman suggested a `dired-jump` action.

**Comment by oantolin** [comment]
I'm not sure there's much advantage in having a flyspell error target, @sheijk, what action other than `ispell-word` would make sense for it? If that's the only one, it's better to press `M-$` directly.

**Comment by jaor** [comment]
On Tue, Feb 09 2021, Omar Antolín Camarena wrote:

> If that's the only one, it's better to press M-$ directly.

hmm, i'm not sure that's the case.  if one uses embark often, it's one
less keybinding to memorize (assuming one remembers there's an
associated action).  of course, M-$ is one keypress less if one has to
choose the action, but couldn't embark be configure to just run the
action when there's only one? (maybe that's already the case)

**Comment by oantolin** [comment]
> but couldn't embark be configure to just run the action when there's only one? (maybe that's already the case)

There is never only one possible action, @jaor. You can always run any command you want as an action (just use `M-x` after `embark-act`, if you are using the keymap prompter, or type the command name if you are using the completing-read prompter). Of course you probably meant, "if the relevant keymap has only one binding", but since they all inherit from `embark-general-map`, there always a bunch of actions available, including saving in the kill ring, inserting in the buffer, etc. OK, but maybe Embark could autorun an action if there is only one action bound in the keymap *that is not in* `embark-general-map`. Well, that could be done, but I wouldn't want to, since you'd lose the ability to use `M-x` for actions in that case.

**Comment by oantolin** [comment]
But your point about learning one fewer keybinding is a good point, @joar, because embark can remind you of the keybinding.

**Comment by minad** [comment]
Yes, having this single entry point context-dependent keybinding is is a big plus of Embark, in particular since the keyboard is already so crowded with Emacs default bindings! It seems you have to promote this `embark-act M-x` feature more heavily :laughing: I would also not like to have a special casing to save keystrokes if only one a single non-general binding exists.

**Comment by jaor** [comment]
On Tue, Feb 09 2021, Omar Antolín Camarena wrote:

> There is never only one possible action, @jaor.

that's true.  what would you think about having the ability of marking
an action as default and having an embark-act-default (or some better
name) command? 

the mark could be added by the embark-define-keymap macro, and maybe use
the convention that the first action in the specific keymap is the
default, so that nothing new (except the new command) needs to be
learned.

**Comment by hmelman** [comment]
Can I suggest adding `hi-lock-face-symbol-at-point` to `embark-symbol-map`.  I don't use it much myself now that I've found `symbol-overlay.el` but the concept is good.

**Comment by oantolin** [comment]
@joar

The current convention is that you bind the "default default action" to `RET` in your type's keymap. The actual default action is given in order of priority by:

- looking up the type in `embark-default-action-overrides`
- the command that opened the minibuffer, if you are in a minibuffer
- the command bound to `RET` in the appropriate action keymap.

I guess `<embark-act> RET` is still at least two keypresses, maybe it would make sense to define a command that just runs the default action.

**Comment by hmelman** [comment]
> what would you think about having the ability of marking an action as default and having an embark-act-default (or some better
name) command?

I did basically this in an old package mouseme.el (which I think is still shipped with ess), it was similar to embark but was a mouse-based menu that figured out thing-at-click and presented a menu of commands.  My favorite use case was a regexp that recognized how we typed  our homegrown bug database numbers, (e.g, PR12234) and set the default action to do a websearch in our bug system but other actions to or copy the text, etc.  At the time I was using emacs for coding and email and reading newsgroups so this worked in any context I saw a bug number and was great.

**Comment by oantolin** [comment]
That sounds very cool, @hmelman, and it's also exactly the type of thing I hope people start using Embark for. Well, people have started, for example @joar has some code to recognize video URLs. (Maybe you want to bind `jao-play-video-url` to `RET` instead of `p`, to indicate it's the default?)

Also GNU Hyperbole's action key does this kind of thing and it comes preconfigured with a large number of recognizers for things, including, I believe Emacs bug numbers.

**Comment by hmelman** [comment]
I've looked at hyperbole a few times but can't get passed the documentation and the idea that it seems like a hodgepodge of stuff much of which I'm not interested in.  But yes I think the core idea is similar I wish it would take the architecture approach of selectrum/consult/embark/marginalia and keep things independent.

But what I didn't state explicitly is that I do like a default action and like the idea of having another command that could typically be bound to the same key with an additional modifier or could use a universal argument or something quick to act as basically the "I feel lucky" button.

**Comment by oantolin** [comment]
> another command that could typically be bound to the same key with an additional modifier

Think of `RET` as a postfix modifier. 😛 So if you bind `embark-act` to, say, `C-,`, then `C-, RET` is your "I'm feeling lucky" button.

**Comment by minad** [comment]
> Think of RET as a postfix modifier. stuck_out_tongue So if you bind embark-act to, say, C-,, then C-, RET is your "I'm feeling lucky" button.

Note that you could easily write a command which wraps this exact behavior and saves you the two keypresses. So you could really put it onto a single button.

**Comment by oantolin** [comment]
> > Think of RET as a postfix modifier. stuck_out_tongue So if you bind embark-act to, say, C-,, then C-, RET is your "I'm feeling lucky" button.
> 
> Note that you could easily write a command which wraps this exact behavior and saves you the two keypresses. So you could really put it onto a single button.

```elisp
(defun embark-default-action ()
  (interactive)
  (pcase-let ((`(,type . ,target) (embark--target)))
    (embark--act (embark--default-action type) target)))
```

**Comment by hmelman** [comment]
I really have to internalize `pcase-let` and use it more, the destructing binding is really nice.  I do it in python, but haven't adopted it since it's come to elisp.

**Comment by oantolin** [comment]
I believe `pcase` was added to Emacs about 8 or 9 years ago, @hmelman. 😛

**Comment by minad** [comment]
> I believe pcase was added to Emacs about 8 or 9 years ago, @hmelman. stuck_out_tongue

It is still broken - it does not get dynamic binding right, due to some invalid optimization on Emacs 28. The tendency seems to be to just disallow dynamic bindings, I disagree with that :-P See https://github.com/minad/consult/issues/108.

**Comment by jaor** [comment]
On Tue, Feb 09 2021, Omar Antolín Camarena wrote:

> (defun embark-default-action ()
>   (interactive)
>   (pcase-let ((`(,type . ,target) (embark--target)))
>     (embark--act (embark--default-action type) target)))

that's exactly what i was looking for, thanks (fwiw, i think it's a
function worth adding to embark proper, or perhaps to its
documentation).

**Comment by hmelman** [comment]
Well maybe in another 8 or 9 years `pcase` will work correctly 🤣

**Comment by oantolin** [comment]
I added `hi-lock-face-symbol-at-point` to the identifier and symbol maps, and added the `embark-default-action` command.

**Comment by minad** [comment]
@hmelman @oantolin Instead of `highlight-symbol-at-point` would it make sense to add a toggler? I thought `highlight-symbol-at-point` does that out of the box, but unfortunately it does not.

~~~ elisp
(defun embark-highlight-symbol ()
  (interactive)
  (let ((regexp (hi-lock-regexp-okay
		  (find-tag-default-as-symbol-regexp))))
    (if (assoc regexp hi-lock-interactive-patterns)
        (hi-lock-unface-buffer regexp)
      (let* ((hi-lock-auto-select-face t)
	     (face (hi-lock-read-face-name)))
        (unless (facep face)
          (setq face 'hi-yellow))
        (unless hi-lock-mode
          (hi-lock-mode 1))
        (hi-lock-set-pattern regexp face)))))
~~~

**Comment by oantolin** [comment]
I like how `highlight-symbol-at-point` works, @minad, and find it better than toggling. It highlights successive symbols in different colors! And a call to `hi-lock-mode` turns off all the various symbols you have highlighted.

**Comment by oantolin** [comment]
Oh, you mean it should turn off the highlighting *only if you call it on the same symbol again* and you still get the multiple symbols in multiple colors if you call it on *different symbols*. (Sorry, I hadn't read the code carefully.) Yes, I do like this.

**Comment by minad** [comment]
Okay, but how do you untoggle specific symbols, via explicit selection in `hi-lock-unface-buffer`? My version highlights the current symbol and if you highlight it again, the highlighting goes away. This sounds to me more useful than the idempotent behavior of `highlight-symbol-at-point` and fits better with the general dwim Embark style.

**Comment by oantolin** [comment]
I think we were writing a comment at the same time, @minad, see the one just above yours.

**Comment by minad** [comment]
Okay cool! The simpler version I wrote above works well and also does not require you to `(require 'hi-lock)` in contrast to the complicated version.

**Comment by oantolin** [comment]
I'll add your function then, maybe I'll call it `embark-toggle-symbol-highlighting` or something like that, so that people who are familiar with `highlight-symbol-at-point` immediately know why we rolled our own instead of just using `highlight-symbol-at-point`.

**Comment by minad** [comment]
In the name of the poor which-key users, I would suggest a shorter name :sweat_smile:

**Comment by oantolin** [comment]
`embark-toggle-highlight`?

**Comment by minad** [comment]
I think `highlight-symbol-at-point` is not widely used/known, and calling it `embark-highlight-symbol` is reasonably clear. This is also the behavior of `symbol-overlay-put` btw, which people seem to use, e.g., https://github.com/minad/consult/issues/215.

**Comment by hmelman** [comment]
> In the name of the poor which-key users, I would suggest a shorter name 😅

I suggest using which-key's support of custom string names embedded in the keymap, this way the command can be embark-toggle-highlight-symbol and the prompt can be "highlight symbol toggle".  I think they're supported by base emacs though I'm not sure how many versions back that's true (maybe it's just reusing the menubar support?).  Maybe you can add support for them for non-which-key users?

**Comment by minad** [comment]
> I suggest using which-key's support of custom string names embedded in the keymap, this way the command can be embark-toggle-highlight-symbol and the prompt can be "highlight symbol toggle"

It is not that easy, which-key does not support that out of the box yet. There is an open PR for menu items. Right now, which-key allows to configure string replacements. But I am still not in favor of an overly verbose name, like embark-toggle-symbol-highlight.

**Comment by hmelman** [comment]
Cool.  You seem to consistently be 3-4 weeks ahead of all my suggestions 😉

**Comment by minad** [comment]
I would find it a bit embarkier if there wouldn't be a `embark-default-action-overrides` but a `[default]` key instead in the keymap. Also why is there an `embark-export` command and an `embark-exporters-alist` after all? Why not bind `embark-export-dired` directly to the export key in the file keymap and so on. The default binding of the export key in the general map could be `embark-collect-snapshot` then. This would be embarkier too. However there is some logic in `embark-export` which checks that candidates are of the same type, transforms candidates and so. So maybe this is not such a good idea. I just thought it could be nice to unify more settings within the Embark keymaps :shrug:

**Comment by indigoviolet** [comment]
Is this the right issue to suggest an `org-insert-link` action for the new `consult-org-heading` items from https://github.com/minad/consult/pull/276?

**Comment by oantolin** [comment]
> Is this the right issue to suggest an `org-insert-link` action for the new `consult-org-heading` items from [minad/consult#276](https://github.com/minad/consult/pull/276)?

Yep. This could go into the embark-consult package (hosted at this same repo). While we're at it, any other actions?

**Comment by astoff** [comment]
> While we're at it, any other actions?

Clocking in would be handy. And since a dedicated command for clocking in would be nice (even though I couldn't convince @minad to include one yet), just jumping to the heading could make sense as an action as well.

**Comment by minad** [comment]
I like the idea of these hyperbole buttons. Embark already supports `embark-target-url-at-point` and `embark-target-file-at-point` via ffap. What about recognizing org links *everywhere*? The org link syntax is sufficiently unique with its double brackets and extensible.

* [[link]]
* [[link][desc]]

**Comment by minad** [comment]
Something like this works in case one wants to use Embark for this purpose. But the benefits are small in contrast to binding `org-open-at-point-global` directly.

~~~ elisp
(defun embark-target-org-link-at-point ()
  (when (org-in-regexp org-link-any-re)
    (cons 'org-link (match-string-no-properties 0))))

(embark-define-keymap embark-org-link-map
  "Keymap for org links."
  ("RET" org-open-link-from-string))

(add-to-list 'embark-target-finders #'embark-target-org-link-at-point)
(add-to-list 'embark-keymap-alist '(org-link . embark-org-link-map))
~~~

**Comment by hmelman** [comment]
As a markdown user, I wonder if this could be generalized to recognize links (a link and its description) in a variety of modes.
- org: [[link][desc]]
- markdown: \[desc](link)
- wikis: [link desc]
- html: \<a href="link">desc\</a>

It's probably only useful if there are a sufficient number of useful actions on these.  In the reverse direction, I have a single global command that inserts my browser's current page info as link in an emacs buffer, formatted correctly for the mode (including as a plain link in programming modes, mostly for comments).

**Comment by minad** [comment]
@hmelman 

> As a markdown user, I wonder if this could be generalized to recognize links (a link and its description) in a variety of modes.

Then maybe Hyperbole buttons are a better fit for you, which claims to recognizes many things implicitly. I don't think a generalized link type is a reasonable addition to Embark. Of course you can define a regexp which recognizes all kinds of links and then invent a new link open command which can handle all these link types. Alternatively you can add embark target finders for all different link types and define separate key maps for each type, dispatching to the corresponding command.

> It's probably only useful if there are a sufficient number of useful actions on these. 

Agree. There are not many useful actions to add to an `embark-org-link-map`. One advantage is that context sensitive bindings are collected under the single `embark-act` binding.

> In the reverse direction, I have a single global command that inserts my browser's current page info as link in an emacs buffer, formatted correctly for the mode (including as a plain link in programming modes, mostly for comments).

There is also `org-insert-link-global` which would do the insertion operation globally.

**Comment by minad** [comment]
@hmelman For Markdown the same approach as above would work obviously:

~~~ elisp
(defun embark-target-markdown-link-at-point ()
  (when (markdown-detect-link...)
    (cons 'markdown-link (match-string-no-properties 0))))

(embark-define-keymap embark-markdown-link-map
  "Keymap for markdown links."
  ("RET" markdown-open-link...))
~~~

However it won't be as useful as org since afaik the markdown links are not as extensible, at least not in its standard definition. You can of course define your own markdown link types `[desc](elisp:...)` which will work in your personal extension of markdown.

**Comment by hmelman** [comment]
I've looked a hyperbole a few times but was never able to get past it's verbose documentation.  I know there's renewed work on it and I'm looking forward to trying again.  

I've dabbled with org but since I also use iOS devices, markdown has been a more portable format for me.  I know that org items have lots of different properties and so consult narrowing based on that is great and doesn't apply to markdown.  But this has got me thinking that a `consult-markdown-links` might be useful to me or a `consult-html-links`.  So now I'm wondering what other actions would be useful on these links and if they generalize?  

Certainly jump to and insert actions are useful. Open in (a variety of) browsers could be useful too.  Copying to the system clipboard as a hyperlink could be useful too (in a way so that it could be pasted into other places as a usable link).  This might be merely academic but it might be nice to select a link in one buffer (say html) and insert into another (say org or markdown).

I'm just thinking out loud here, I don't know if this is useful or if it's more suitable to embark or another package or just wiki idea.

>  I don't think a generalized link type is a reasonable addition to Embark

Why do you think this?

**Comment by minad** [comment]
> > I don't think a generalized link type is a reasonable addition to Embark
>
> Why do you think this?

The embark target types seem more concrete. There is a type 'file, 'url and so on, but not something like 'general-link-string. I think 'org-link would be a good fit if there would be more special actions besides the default open action. Of course there are the general actions.

> I've looked a hyperbole a few times but was never able to get past it's verbose documentation. I know there's renewed work on it and I'm looking forward to trying again.

Yes, my main criticism is that it integrates different concepts which don't seem very related (buttons, outliner, contacts, window management?!) and it seems to invent its own menu ui. Org can also be criticized since it does everything. But org has the extensible markup and outliner language at its core, and all the other features revolve around that.

> I've dabbled with org but since I also use iOS devices, markdown has been a more portable format for me.

I agree that markdown is a more portable format - we all use it a lot since it is just vastly popular in the web. But I've always found it far too limited. Initially I found some parts of the org syntax a bit ugly (e.g., `#+begin_src` blocks), but one gets accustomed to that and the extensibility is a significant advantage. I've also tried asciidoc for a while, which is like an extended markdown, but it is too niche and not Emacs friendly.

**Comment by hmelman** [comment]
So I have the above working in a rough form.  I have a `consult-link` based on `consult-outline` that finds all the links in the buffer.  They're rendered in the minibuffer as just the description text (I'm not sure how).  I created a link type for embark with actions for insert in format of current buffer and open in browse-url and eww.  An embark collect snapshot is nice and it's nice to grab a link from earlier in a buffer and insert it at point.  So far it's pretty cute.  I'll keep playing with it to see it if it's actually useful. :)

**Comment by minad** [comment]
Why do you need a `consult-link` command? When I was initially talking about the links I thought about Embark actions at point, not Embark actions in the minibuffer.

**Comment by hmelman** [comment]
Because it's useful in a buffer full of links 😄

**Comment by oantolin** [comment]
@protesilaos also has a `consult-link`-type command for links in an eww buffer: `prot-eww-visit-url-on-page`.

**Comment by protesilaos** [comment]
> @protesilaos also has a consult-link-type command for links in an eww buffer: prot-eww-visit-url-on-page.

I also have those as well:

```elisp
(defvar prot-common-url-regexp
  (concat
   "\\b\\(\\(www\\.\\|\\(s?https?\\|ftp\\|file\\|gopher\\|"
   "nntp\\|news\\|telnet\\|wais\\|mailto\\|info\\):\\)"
   "\\(//[-a-z0-9_.]+:[0-9]*\\)?"
   (let ((chars "-a-z0-9_=#$@~%&*+\\/[:word:]")
	     (punct "!?:;.,"))
     (concat
      "\\(?:"
      ;; Match paired parentheses, e.g. in Wikipedia URLs:
      ;; http://thread.gmane.org/47B4E3B2.3050402@gmail.com
      "[" chars punct "]+" "(" "[" chars punct "]+" ")"
      "\\(?:" "[" chars punct "]+" "[" chars "]" "\\)?"
      "\\|"
      "[" chars punct "]+" "[" chars "]"
      "\\)"))
   "\\)")
  "Regular expression that matches URLs.
Copy of variable `browse-url-button-regexp'.")

;; [...]

;;;###autoload
(defun prot-search-occur-urls ()
  "Produce buttonised list of all URLs in the current buffer."
  (interactive)
  (add-hook 'occur-hook #'goto-address-mode)
  (occur prot-common-url-regexp "\\&")
  (remove-hook 'occur-hook #'goto-address-mode))

;;;###autoload
(defun prot-search-occur-browse-url ()
  "Point browser at a URL in the buffer using completion.
Which web browser to use depends on the value of the variable
`browse-url-browser-function'.

Also see `prot-search-occur-urls'."
  (interactive)
  (let ((matches nil))
    (save-excursion
      (goto-char (point-min))
      (while (search-forward-regexp prot-common-url-regexp nil t)
        (push (match-string-no-properties 0) matches)))
    (funcall browse-url-browser-function
             (completing-read "Browse URL: " matches nil t))))
```

I normally use those with my plain text files (no, I am still not an Org fan) where the URL is just written as `LINK` or `<LINK>`, so they do not do fancy stuff like showing an arbitrary title or some meta datum while actually pointing to the URL.

**Comment by oantolin** [comment]
I added a "textual representation of a keyboard macro" target together with the obvious kmacro actions (run, save on kmacro ring, name or bind) to the wiki, [here](https://github.com/oantolin/embark/wiki/Additional-Actions#actions-for-textual-representations-of-keyboard-macros). I used Hyperbole's nice brace notation for kmacros.

**Comment by minad** [comment]
> I added a "textual representation of a keyboard macro" target...

That's nice! Thanks! My problem with the Hyperbole notation is that it is a little too ambiguous to activate it everywhere. See also the discussion https://www.reddit.com/r/emacs/comments/n3oqle/gnu_hyperbole_v8_rocket_through_your_information/gws669v/. I hope I've not been unfairly critical there. :)

**Comment by oantolin** [comment]
> My problem with the Hyperbole notation is that it is a little too ambiguous to activate it everywhere.

Is it, the braces? What problems do you foresee? I don't really imagine any problems with the braces, but I think I might prefer Org links as you explained in that reddit thread: it's a nice explicit syntax, but still short. Even though I wrote the braces target finder and some semi-reasonable actions, I didn't even turn it on by default in my personal configuration.

I wonder which approach is better for running the key sequences: `kmacro-call-macro` or `unread-command-events`? I didn't attempt any comparison of the two approaches, nor did I experiment a lot.

**Comment by minad** [comment]
> I don't really imagine any problems with the braces

Probably you are right. And we are still explicitly triggering the actions. So no problem really. This would matter more if there is some automatic stuff going on.

> I wonder which approach is better for running the key sequences: kmacro-call-macro or unread-command-events? I didn't attempt any comparison of the two approaches, nor did I experiment a lot.

I think for your Embark actions the kmacro is the correct approach, since the key sequences must be proper kmacros. If you use unread-command-events you can also write improper/prefix "macros" like {M-x} or {C-x}.

**Comment by oantolin** [comment]
I had not realized you cannot stop recording right after `M-x`! If you try `<f3> M-x <f4>` you get "Ignore empty macro". This seems like an unfair limitation of keyboard macros.

**Comment by minad** [comment]
It seems macros are supposed to end somehow in a state without pending events. One can then translate the macro representation to a sequence of function calls. With such prefix "macros" you cannot do that. I also wondered about this - it was probably some very early design decision.

I imagine the bug report discussion if someone reports this as a "bug", like 40 years after the inception of Emacs :D

**Comment by minad** [comment]
What I wanted to add - for a text based representation there is still some value in having these "prefix macros", e.g., when writing some documentation: Press {C-x r j} to jump to a register...

**Comment by oantolin** [comment]
I wonder if Hyperbole supports these incomplete macros?

**Comment by minad** [comment]
There is a place to ask, this just came up https://www.reddit.com/r/emacs/comments/nirwpk/my_understanding_of_gnu_hyperbole/ :)

**Comment by oantolin** [comment]
Well, it seemed easier to `{M-x try RET hyperbole RET}`, and then check with `{C-x r j}`. It does support them.

**Comment by minad** [comment]
I've never tried try. The problem is that I cannot try try without try.

**Comment by oantolin** [comment]
I did not know keyboard macros had this limitation. I'm pretty upset now. (🤣) I wonder if I should switch to `unread-command-events` for the run action (keeping the limitation for the other actions... in this little package I probably won't use...)

> I've never tried try. The problem is that I cannot try try without try.

It requires a leap of faith. 😄 I think it's worth installing. All it does is install to `/tmp` so that if you restart Emacs the packages you `try`ed are gone.

**Comment by Zetagon** [comment]
I have created a pretty general target finder that might be useful in embark (or `embark-org`/`org-embark`) [here](https://github.com/Zetagon/literate-dotfiles/blob/master/config.org#org).  It can be used to implement target finders for any org element (I have already used it for org links). I think it would be pretty useful to have as upstream as possible because it makes it easier to customize.

What do you guys think?

**Comment by minad** [comment]
@Zetagon Great. I wished for something like this before. If you ask me, something like this is a good fit for Embark, but the decision is of course @oantolin's.

**Comment by Zetagon** [comment]
> EDIT: @Zetagon I took a closer look at your code and I don't understand what you did there. I thought you added target finders for some specific org elements with appropriate actions. Org-ref specific actions cannot be added to Embark directly due to the third party dependencies.

@minad The org-ref part is more of an example of how a user (me) can add their own actions specific to a certain element. `my/embark-target-org-context' allows the user to add their own finders for a specific org element. In this case I've added `my/embark-target-org-link' for links. `my/embark-target-org-link' is also a general finder, the user can add their own specific actions depending on the link type. In this example I, the user, have added actions that work specifically for org-ref cite and ref links.

Hope this made it clearer!

**Comment by minad** [comment]
@Zetagon why ist there a need for separate org target finders? Why not add them to the list of usual embark target finders?

**Comment by Zetagon** [comment]
> Why not add them to the list of usual embark target finders?
@minad The idea came to me when I was implementing a target finder for org-ref links. I first used a simple approach with regexps, but I quickly realized that a much better alternative is to use org-mode's own parser. It makes it much easier to add new action types for certain links, and it generalized pretty well to any org element.

**Comment by minad** [comment]
@Zetagon I see. You do the parsing once in the embark-registered classifier and then pass the org element to the org-specific classifier. This makes sense. On the other hand one could also hard code all possible org elements in the embark-classifier. There are not that many (org-headline, org-link, org-block, org-table, org-drawer, ...)? Furthermore it only makes sense to add the ones for which we have some actions to add. This is the first - identifying the actions which are needed?

See https://github.com/oantolin/embark/issues/95#issuecomment-758246901 by @sheijk.

**Comment by Zetagon** [comment]
@minad I guess hard-coding could work pretty well since new elements aren't easily added (to my knowledge at least). But I personally like the idea of easily adding a target for, say a bold element or a clocked entry, without having to worry about conflicting with other finders (assuming of course, that embark doesn't include a target for every single org element). 

Links are extensible and many packages add their own link types, and for some link types separate actions makes sense (org-ref being just an example here). I think my approach of having a central target finder would make sense here, because it makes it easier to add new link types.

I can understand that having a central target finder like this might be going against the idea of embark, and add unnecessary complexity. I do still think this is useful, so maybe it could be a separate package.

**Comment by bdarcus** [comment]
> I think your suggested org targets and actions are very good and that a separate embark-org (or org-embark?, how does that work?) package with them would be reasonable

@sheijk - what's the status of this idea (of something like a `org-embark`)?

In addition to supporting your list of possible target finders for org-mode, I was wanting to recommend one for the forthcoming `org-cite` feature, and its citation and citation-reference objects.

**Comment by sheijk** [comment]
@bdarcus org-menu is only a transient based interface for org-mode. I've just listed a few things which would be good candidates for embark action, too. I'm not working on anything like that

**Comment by minad** [comment]
@oantolin What about sexp and defun targets at point? The sexp at point target could be active when the point is located on the parentheses. defun at point could be active when the point is located on "(defun" or on the closing ")".

~~~ elisp
;; Only active when cursor is located on parentheses
(defun embark-target-sexp-at-point ()
  "Target sexp at point."
  (save-excursion
    (when (eq (char-after) ?\))
      (forward-char 1))
    (when-let (bounds (bounds-of-thing-at-point 'sexp))
      (and (eq (char-after (car bounds)) ?\()
           (cons 'sexp bounds)))))
~~~

**Comment by minad** [comment]
See also https://github.com/jyp/cmap/blob/main/cmap.el by @jyp. There are a few more actions at point we can import into Embark.

**Comment by minad** [comment]
My idea regarding sexps is to allow the cursor at the following positions:

~~~ elisp
(sexp)
^    ^^
~~~

The point must be located at a parenthesis. The commands should be normalized with respect to the point position, so that all the commands work in all the scenarios. And for defun I would require a location at "(defun" or ")". Otherwise one gets too many target matches, conflicting with the other target finders.

**Comment by minad** [comment]
Another idea would be to have different target types at a point and merge the keymaps. But this is probably an unnecessary complication and introduces the keymap shadowing problem. This has been discussed before in #92.

**Comment by oantolin** [comment]
Yeah, normalizing the cursor position sounds like the easiest approach. We make a list of the desired sexp actions and if `last-sexp` is more popular than pre `sexp`, and arrange for the least popular type to have point in the right place. This can be done by writing little wrappers, of course, but maybe also in `embark-pre-action-hook`.

**Comment by hmelman** [comment]
sexp targets should work at points that the sexp navigation commands leave you (C-M-f C-M-b C-M-u C-M-d C-M-n C-M-p).

**Comment by oantolin** [comment]
> Another idea would be to have different target types at a point and merge the keymaps. But this is probably an unnecessary complication and introduces the keymap shadowing problem. This has been discussed before in #92.

I think this was exactly the motivation of the cmap package.

**Comment by oantolin** [comment]
> sexp targets should work at points that the sexp navigation commands leave you (C-M-f C-M-b C-M-u C-M-d C-M-n C-M-p).

So that's @minad's list of positions plus right inside the opening parenthesis, right?

**Comment by hmelman** [comment]
> > sexp targets should work at points that the sexp navigation commands leave you (C-M-f C-M-b C-M-u C-M-d C-M-n C-M-p).
> 
> So that's @minad's list of positions plus right inside the opening parenthesis, right?

Seems to be.

**Comment by minad** [comment]
> I think this was exactly the motivation of the cmap package.

But how does cmap solve this issue? I don't think it actually does - it just avoids shadowing in the keymaps?

         

> So that's @minad's list of positions plus right inside the opening parenthesis, right?
> > Seems to be.

No, right inside the opening parenthesis is too conflicting, since there you have a symbol with the symbol actions at point. I excluded that position intentionally.

**Comment by oantolin** [comment]
> But how does cmap solve this issue? I don't think it actually does - it just avoids shadowing in the keymaps?

I think so, yes. If I recall correctly, jyp thought merging keymaps was enough, that users should take care to only shadow when they didn't mind the shadowing. I wanted a more robust way to deal with the problem, but as usual procrastinated in the face of multiple UI design choices and didn't implement anything (well, no, I did make a branch with the "merge and hope for the best" approach). (I think jyp also though Embark's calling convention was overly complex and wanted something simpler.)

> No, right inside the opening parenthesis is too conflicting, since there you have a symbol with the symbol actions at point. I excluded that position intentionally.

I agree.

**Comment by minad** [comment]
@oantolin Regarding dap/cmap - I think the motivation for the package is mainly to have something conceptually simpler. You may remember that I initially wondered why Embark conflates actions at point with for example minibuffer candidate actions. Now I think this is a nice unification. But if you are only interested in actions at point, you can use a simpler implementation. The action can always obtain the target from the point and one avoids the minibuffer injection calling convention from Embark.

**Comment by oantolin** [comment]
Seems like we are in agreement.

**Comment by minad** [comment]
:laughing:

**Comment by Zetagon** [comment]
@oantolin Any comment on this from earlier in this thread? https://github.com/oantolin/embark/issues/95#issuecomment-855268596

**Comment by minad** [comment]
@Zetagon Related to the current discussion - the aforementioned cmap package implements a set of useful org actions at point and a set of org target finders. You may want to take a look what exists there.

**Comment by oantolin** [comment]
I'm sorry @Zetagon, I haven't really had a chance to think about this.

I do like the idea of having a target finder for org elements, but I think deciding what that target finder should return is a little tricky and requires contemplating what actions one would want. The two natural choices for it are to return the org element as a string, and to return a non-string (like the region target finder which just returns the symbol `<region>` if there is an active region). The point is that if the target finder returns a string, that string will be injected at the first minibuffer prompt that the action produces; but if the target is not a string, nothing is injected.

So you need to look at the likely actions and see how they actually work. In the org case most probably don't take the org-element as a string, but rather act on the thing at point. I also find it hard to decide whether the layered approach you took, with extra configuration variables for elements types and link types is really necessary. I haven't forgotten this, but haven't had time to think about it yet.

**Comment by Zetagon** [comment]
@oantolin 
> I'm sorry @Zetagon, I haven't really had a chance to think about this.

Yeah that is completely understandable. I just didn't want it to drown in the discussion afterwards :slightly_smiling_face:  Do take your time, I will just find something else to play with! 

> 
> I do like the idea of having a target finder for org elements, but I think deciding what that target finder should return is a little tricky and requires contemplating what actions one would want. 

In my current implementation each element has its own target finder so what we return can depend on the target. For example the link target returns the link string, but that doesn't stop target finders for other elements to return, say a symbol or the parsed AST. Is there an advantage having the same return type for all elements I'm missing?

> I also find it hard to decide whether the layered approach you took, with extra configuration variables for elements types and link types is really necessary. I haven't forgotten this, but haven't had time to think about it yet.

I think there is a greater benefit to having configuration variables for links than for element types at least, because users can add their own link types.

**Comment by astoff** [comment]
> ```
> (sexp)
> ^    ^^
> ```

In my opinion the second of those positions should target the symbol `sexp` inside the sexp.

> sexp targets should work at points that the sexp navigation commands leave you (C-M-f C-M-b C-M-u C-M-d C-M-n C-M-p).

An alternative would be to follow `show-paren-mode`'s example.

**Comment by oantolin** [comment]
Oh, right! I fully agree with @astoff: `show-paren-mode` has it right: target the sexp only right before an opening parenthesis or right after a closing parenthesis. And as @astoff also said, right before the closing parenthesis the target should be the symbol to the left, which is what currently happens.

**Comment by oantolin** [comment]
@Zetagon

> Yeah that is completely understandable. I just didn't want it to drown in the discussion afterwards 🙂 Do take your time, I will just find something else to play with!

Thanks for being understanding. We should probably create a separate issue for this, since as you say it could get drowned out here.
 
> In my current implementation each element has its own target finder so what we return can depend on the target. For example the link target returns the link string, but that doesn't stop target finders for other elements to return, say a symbol or the parsed AST. Is there an advantage having the same return type for all elements I'm missing?

Oh, no, there's no advantage in all of them returning the same type, and that is not what I meant. I meant we should consider for _each type_ what to return.
 
> I think there is a greater benefit to having configuration variables for links than for element types at least, because users can add their own link types.

Yes, I agree.

**Comment by hmelman** [comment]
> Oh, right! I fully agree with @astoff: `show-paren-mode` has it right: target the sexp only right before an opening parenthesis or right after a closing parenthesis. And as @astoff also said, right before the closing parenthesis the target should be the symbol to the left, which is what currently happens.

Agreed.  Is this only for lisp modes or should this work in other programming modes that define sexps in some way (I think most do as expressions)?  The mark, kill and indent actions would apply.  Maybe not eval or expand though a send to interpreter might.

**Comment by minad** [comment]
Such a sexp target finder is quite simple, even simpler than the one I proposed before:

~~~ elisp
(defun embark-target-sexp-at-point ()
  "Target sexp at point."
  (when (and (or (eq (char-after) ?\()
                 (eq (char-before) ?\)))
             (thing-at-point 'sexp))
    '(sexp <sexp>)))
~~~

**Comment by oantolin** [comment]
@hmelman It would be nice if it worked in other programming modes as well, but in my experience the definition of sexp often isn't good enough. Try `mark-sexp` in different places and you'll see what I mean. It does mark a block in braces, but that's about the only thing it does right. In C, for example, `if (a==b) { printf("Oh no!\n"); }` it thinks `if` is an expression...

**Comment by oantolin** [comment]
Given the state of sexps in other modes, I think a minor tweak of @minad's target finder to use syntax class `(` instead of the literal character `(` is probably as good as it gets without excessive work.

Oh, maybe syntax class `(` and `"` should be considered.

**Comment by minad** [comment]
@oantolin I am not sure about " since this may once again clash with other finders. file names? If we use only the parentheses we are at least safe that nothing bad happens.

**Comment by minad** [comment]
But generally one could add additional target finders which are programming language specific. Or maybe it is even possible to do a bit better with tree sitter. However I think such actions should be maintained in third party packages.

**Comment by oantolin** [comment]
I think file names in strings should be safe. Given `"~/a/path/"` ffap only recognizes the file if you are inside the string, this leaves the positions just outside the string on either side available for treating the string as an s-expression.

**Comment by hmelman** [comment]
I have the following in a `markdown-mode` buffer.

    - [Maracaibo: The Uprising](https://boardgamegeek.com/boardgameexpansion/338013/maracaibo-uprising)

I wanted to use the link text, "Maracaibo: The Uprising", in an external app (a browser search).  I positioned point inside the word "Maracaibo" and could invoke `expand-region` twice and use `M-w` to copy the desired text but I wanted to use embark.

I didn't think it would work but invoking `embark-act` twice highlighted the desired text as an expression, almost.  It included the open and close brackets because that's how sexps work in text mode.

I'd love if embark could act on this text as a target in `markdown-mode`.  I suspect it would be useful in some other modes like org or html.  I'm not sure if expression is the right target to use.  `expand-region` works by selecting first `er/mark-inside-pairs` and then `er/mark-outside-pairs`.  I could see that being useful in `prog-modes` too, at least in comments (it also has functions for inside/outside quotes).

Previously in this issue we talked about targets for links, but this is different as I don't want the url or the whole link, I think I want the text marked by a delimiter surrounding point which seems generally useful, but maybe someone has other ideas.

**Comment by minad** [comment]
What about new targets button and widget? We can use `button-at` and `widget-at` to find them. There are not many actions, but Emacs 28 introduced `button-describe` and `widget-describe`.

**Comment by oantolin** [comment]
To act inside the delimiters of an s-expression, @hmelman, you could use this (which takes advantage of shiny new infrastructure I just added in f741dab05b09beb18e0a7e87f5b80ea462ca44a2):

    (defun embark-mark-inside ()
      "Mark inside an s-expression, excluding the delimiters."
      (interactive)
      (forward-char)
      (push-mark)
      (up-list)
      (backward-char)
      (activate-mark))

    (add-to-list 'embark-repeat-actions '(embark-mark-inside . region))

    (add-to-list 'embark-pre-action-hooks
                 '(embark-mark-inside embark--beginning-of-target))

    (define-key embark-expression-map "I" #'embark-mark-inside)
    
The new thing added in that commit was the ability to add pairs of the form `(action . next-target-type)` to embark-repeat-actions. Previously, at which point in the target cycle you started the next action was hard-coded.

I'm not sure whether to add that to Embark (and wouldn't mind hearing some opinions), but if not, it can certainly go on the wiki.

**Comment by minad** [comment]
Looks a bit special to me, but we also have hash function actions so why not? I am not sure where to draw the line. ;)

**Comment by oantolin** [comment]
Oh, I should probably check the first and last character really are delimiters, because that's not always true of sexps in non-lisp prog modes. 🙄

**Comment by minad** [comment]
> Oh, I should probably check the first and last character really are delimiters, because that's not always true of sexps in non-lisp prog modes.

Hmm, maybe this is a line I would draw. If an action gets too complicated it should not be added, since then there is no clear line on where to stop. While the hash actions are obscure they are only trivial wrappers. Of course exceptions can be made for complex actions, which are considered important.

**Comment by pcrama** [comment]
It would be great if EMBARK was made aware of `bug-reference-mode` and picked the URL as a target out of the box.

I have code in my init.el to do that by looking at the overlay category (`bug-reference`) and `(overlay-get o 'bug-reference-url)` of the `(overlays-at (point))` but since I did not sign the FSF papers, I do not dare paste my code here.

**Comment by minad** [comment]
@pcrama Embark already does. See https://github.com/oantolin/embark/blob/592287b7d7393f6f13225b722f550d59cf2eb665/embark.el#L766-L771.

**Comment by pcrama** [comment]
Thanks and sorry for missing it. I still don't know how I initially got the impression it wasn't working but now it works for me.

I don't know how to close this, unfortunately but from my point of view, this issue is resolved.

**Comment by oantolin** [comment]
I've started an embark-org package in this repo, which is one of things discussed long ago in this issue. I open a separate issue for discussion of what to add to embark-org, #502

**Comment by jumper047** [comment]
BTW what is the preferable way to contribute my own finders/collectors? Recently I tinkered with embark and emacs-slack packages and added some actions on slack messages. Seems like it can be useful for someone, and I have the choice - to contribute it to the main embark repo or to go with separate package on my own.
(sorry if it's wrong topic for such questions)
